File: hack/openshift/update_configmap.go
package main

import (
	"flag"
	"fmt"
	"os"
	"strings"

	"gopkg.in/yaml.v3"
)

const (
	defaultAgentPullspec  = "registry.redhat.io/bpfman/bpfman-agent@sha256:4105eb14731e7f41c177f8a5ac0be7447967f4ec3338967e15f398d3147225b3"
	defaultBpfmanPullspec = "registry.redhat.io/bpfman/bpfman@sha256:4c31043d37cd20bb43fcb64d38b9c8cdfb8d1c9317d7d8b24051f9005abd7112"
)

func main() {
	var (
		configMapFile  = flag.String("configmap-file", "", "Path to the ConfigMap file to update")
		agentPullspec  = flag.String("agent-pullspec", defaultAgentPullspec, "bpfman-agent image pullspec")
		bpfmanPullspec = flag.String("bpfman-pullspec", defaultBpfmanPullspec, "bpfman image pullspec")
		outputFile     = flag.String("output", "", "Output file (defaults to input file)")
		dryRun         = flag.Bool("dry-run", false, "Show what would be changed without modifying files")
		help           = flag.Bool("help", false, "Show help message")
	)

	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "Usage: %s [options] <configmap-file> [output-file]\n\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "Update bpfman ConfigMap with Red Hat image references.\n\n")
		fmt.Fprintf(os.Stderr, "Options:\n")
		flag.PrintDefaults()
		fmt.Fprintf(os.Stderr, "\nExamples:\n")
		fmt.Fprintf(os.Stderr, "  %s --configmap-file input.yaml\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "  %s --dry-run input.yaml\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "  %s input.yaml output.yaml\n", os.Args[0])
	}

	flag.Parse()

	if *help {
		flag.Usage()
		os.Exit(0)
	}

	// Sanitize pullspecs by removing leading/trailing whitespace and newlines
	*agentPullspec = strings.TrimSpace(*agentPullspec)
	*bpfmanPullspec = strings.TrimSpace(*bpfmanPullspec)

	// Handle positional arguments
	args := flag.Args()
	if *configMapFile == "" && len(args) > 0 {
		*configMapFile = args[0]
	}
	if *outputFile == "" && len(args) > 1 {
		*outputFile = args[1]
	}

	// Default output to input file if not specified
	if *outputFile == "" {
		*outputFile = *configMapFile
	}

	if *configMapFile == "" {
		fmt.Fprintf(os.Stderr, "Error: ConfigMap file is required\n\n")
		flag.Usage()
		os.Exit(1)
	}

	if *dryRun {
		fmt.Printf("Dry run mode: would update %s\n", *configMapFile)
		if *outputFile != *configMapFile {
			fmt.Printf("Output would be written to: %s\n", *outputFile)
		}
		fmt.Printf("Agent pullspec: %s\n", *agentPullspec)
		fmt.Printf("Bpfman pullspec: %s\n", *bpfmanPullspec)
	}

	fmt.Printf("Updating ConfigMap file: %s\n", *configMapFile)

	content, err := os.ReadFile(*configMapFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file %s: %v\n", *configMapFile, err)
		os.Exit(1)
	}

	contentStr := string(content)

	replacements := map[string]string{
		"quay.io/bpfman/bpfman-agent:latest": fmt.Sprintf("\"%s\"", *agentPullspec),
		"quay.io/bpfman/bpfman:latest":       fmt.Sprintf("\"%s\"", *bpfmanPullspec),
	}

	if *dryRun {
		fmt.Println("\nString replacements that would be made:")
		for old, new := range replacements {
			if strings.Contains(contentStr, old) {
				fmt.Printf("  %s -> %s\n", old, new)
			}
		}
		fmt.Println("\nWould also update ConfigMap data fields.")
		os.Exit(0)
	}

	for old, new := range replacements {
		contentStr = strings.ReplaceAll(contentStr, old, new)
	}

	// Parse YAML for structured modifications
	var configMap map[string]interface{}
	err = yaml.Unmarshal([]byte(contentStr), &configMap)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error parsing YAML: %v\n", err)
		os.Exit(1)
	}

	// Ensure data section exists
	data, ok := configMap["data"].(map[string]interface{})
	if !ok {
		data = make(map[string]interface{})
		configMap["data"] = data
	}

	// Update the data fields
	data["bpfman.agent.image"] = *agentPullspec
	data["bpfman.image"] = *bpfmanPullspec

	output, err := yaml.Marshal(configMap)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error marshalling YAML: %v\n", err)
		os.Exit(1)
	}

	err = os.WriteFile(*outputFile, output, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error writing file %s: %v\n", *outputFile, err)
		os.Exit(1)
	}

	fmt.Printf("ConfigMap file updated successfully: %s\n", *outputFile)
}

File: hack/openshift/update-configmap.py
#!/usr/bin/env python3

"""Update bpfman ConfigMap with Red Hat image references.

This tool transforms upstream bpfman-operator ConfigMap manifests for
Red Hat distribution by replacing upstream image references with Red
Hat registry pullspecs whilst preserving YAML formatting and
structure.
"""

import argparse
import sys
import os
from ruamel.yaml import YAML


def load_yaml_file(file_path):
    yaml = YAML()
    yaml.preserve_quotes = True
    yaml.width = 4096
    try:
        with open(file_path, "r") as f:
            return yaml.load(f)
    except FileNotFoundError:
        print(f"Error: File not found: {file_path}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error loading YAML file {file_path}: {e}", file=sys.stderr)
        sys.exit(1)


def save_yaml_file(file_path, data):
    yaml = YAML()
    yaml.preserve_quotes = True
    yaml.width = 4096
    try:
        with open(file_path, "w") as f:
            yaml.dump(data, f)
    except Exception as e:
        print(f"Error writing YAML file {file_path}: {e}", file=sys.stderr)
        sys.exit(1)


def update_configmap(configmap_file, agent_pullspec, bpfman_pullspec, output_file=None):

    if output_file is None:
        output_file = configmap_file

    print(f"Updating ConfigMap file: {configmap_file}")

    configmap = load_yaml_file(configmap_file)

    with open(configmap_file, "r") as f:
        content = f.read()

    replacements = {
        "quay.io/bpfman/bpfman-agent:latest": f'"{agent_pullspec}"',
        "quay.io/bpfman/bpfman:latest": f'"{bpfman_pullspec}"',
    }

    for old, new in replacements.items():
        content = content.replace(old, new)

    yaml = YAML()
    yaml.preserve_quotes = True
    yaml.width = 4096
    configmap = yaml.load(content)

    if "data" not in configmap:
        configmap["data"] = {}

    configmap["data"]["bpfman.agent.image"] = agent_pullspec
    configmap["data"]["bpfman.image"] = bpfman_pullspec

    save_yaml_file(output_file, configmap)

    print(f"ConfigMap file updated successfully: {output_file}")


def main():
    parser = argparse.ArgumentParser(
        description="Update bpfman ConfigMap with Red Hat image references.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""Examples:
  %(prog)s --configmap-file input.yaml --agent-pullspec registry.redhat.io/bpfman/agent@sha256:... --bpfman-pullspec registry.redhat.io/bpfman/bpfman@sha256:...
  %(prog)s --configmap-file input.yaml --output output.yaml --agent-pullspec registry.redhat.io/bpfman/agent@sha256:... --bpfman-pullspec registry.redhat.io/bpfman/bpfman@sha256:...
""",
    )

    parser.add_argument(
        "--configmap-file", required=True, help="Path to the ConfigMap file to update"
    )
    parser.add_argument(
        "--agent-pullspec", required=True, help="bpfman-agent image pullspec"
    )
    parser.add_argument(
        "--bpfman-pullspec", required=True, help="bpfman image pullspec"
    )
    parser.add_argument("--output", help="Output file (defaults to input file)")

    args = parser.parse_args()

    agent_pullspec = args.agent_pullspec.strip()
    bpfman_pullspec = args.bpfman_pullspec.strip()

    if not agent_pullspec:
        print("Error: Agent pullspec cannot be empty", file=sys.stderr)
        sys.exit(1)

    if not bpfman_pullspec:
        print("Error: Bpfman pullspec cannot be empty", file=sys.stderr)
        sys.exit(1)

    update_configmap(args.configmap_file, agent_pullspec, bpfman_pullspec, args.output)


if __name__ == "__main__":
    main()
